File: library/src/main/java/org/fbme/lib/iec61499/instances/Instance.java
package org.fbme.lib.iec61499.instances;


import org.fbme.lib.iec61499.declarations.Declaration;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public interface Instance {

    @Nullable Instance getParent();

    @NotNull Declaration getDeclaration();

    default @NotNull Instance getRootInstance() {
        Instance root = this;
        Instance parent = this.getParent();
        while (parent != null) {
            root = parent;
            parent = root.getParent();
        }
        return root;
    }

}

File: code/library/src/main/kotlin/org/fbme/lib/iec61499/instances/Instance.kt
package org.fbme.lib.iec61499.instances

import org.fbme.lib.common.Declaration

interface Instance {
    val parent: Instance?
    val declaration: Declaration
    val rootInstance: Instance
        get() {
            var root = this
            var parent = parent
            while (parent != null) {
                root = parent
                parent = root.parent
            }
            return root
        }
}
